pkg/datapath/linux/arp/ping.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package arp
+
+import (
+	"bytes"
+	"errors"
+	"net"
+	"time"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"github.com/vishvananda/netlink"
+)
+
+var (
+	ErrNotImplemented = errors.New("not implemented")
+
+	timeout = 1 * time.Second
+)
+
+var defaultSerializeOpts = gopacket.SerializeOptions{
+	FixLengths:       true,
+	ComputeChecksums: true,
+}
+
+// PingOverLink performs arping request from 'src' IP address to the 'dst' IP address
+// over the link 'link' and returns the hardware address (MAC) of the destination
+func PingOverLink(link netlink.Link, src net.IP, dst net.IP) (net.HardwareAddr, error) {
+	p, err := newPinger(link, src)
+	if err != nil {
+		return nil, err
+	}
+	defer p.close()
+
+	if err := p.setDeadline(time.Now().Add(timeout)); err != nil {
+		return nil, err
+	}
+
+	return p.resolve(dst)
+}
+
+var _ net.Addr = &Addr{}
+
+type Addr struct {
+	net.HardwareAddr
+}
+
+func (a *Addr) Network() string {
+	return "raw"
+}
+
+type pinger struct {
+	c    net.PacketConn
+	ip   net.IP
+	link netlink.Link
+}
+
+func (p *pinger) close() {
+	_ = p.c.Close()
+}
+
+func (p *pinger) setDeadline(t time.Time) error {
+	return p.c.SetDeadline(t)
+}
+
+func (p *pinger) resolve(ip net.IP) (net.HardwareAddr, error) {
+	if err := p.request(ip); err != nil {
+		return nil, err
+	}
+
+	for {
+		resp, err := p.read()
+		if err != nil {
+			return nil, err
+		}
+
+		if resp.Operation != layers.ARPReply || !bytes.Equal(resp.SourceProtAddress, ip.To4()) {
+			continue
+		}
+
+		dstHwAddr := make(net.HardwareAddr, resp.ProtAddressSize)
```

Comment:
I think it supposed to be `resp.HwAddressSize`.

Reply:
Yep, looks like I accidentally added wrong field there.